### PR TITLE
Sort edit commands during batch updating of table and collection views

### PR DESCRIPTION
**Pull request by nguyenhuy**
#684, #693.

@levi, @Adlai-Holler: Please also let me know your thoughts on this PR. Thanks.

- [x] Implement ASChangetSetDataController 
- [x] Failing ASTableViewTests (#758)
- [x] Write documentation for ASChangetSetDataController
- [x] Remove duplicate steps while coalescing changes in _ASHierarchyChangeSet
- [ ] Write tests for ASChangetSetDataController

### CODE REVIEW

**Comment by levi** [comment]
Gonna check this out tonight, @nguyenhuy.

File: AsyncDisplayKit/Private/ASInternalHelpers.h [hunk 1]
```diff
@@ -26,3 +26,7 @@ CGFloat ASCeilPixelValue(CGFloat f);
 CGFloat ASRoundPixelValue(CGFloat f);
 
 ASDISPLAYNODE_EXTERN_C_END
+
+@interface NSIndexPath (ASInverseComparison)
+- (NSComparisonResult)as_inverseCompare:(NSIndexPath *)otherIndexPath;
```
Comment (levi) [inline AsyncDisplayKit/Private/ASInternalHelpers.h]: `asdk_inverseCompare`

File: AsyncDisplayKit/Private/ASInternalHelpers.mm [hunk 2]
```diff
@@ -70,3 +70,12 @@ CGFloat ASRoundPixelValue(CGFloat f)
 {
   return roundf(f * ASScreenScale()) / ASScreenScale();
 }
+
+@implementation NSIndexPath (ASInverseComparison)
+
+- (NSComparisonResult)as_inverseCompare:(NSIndexPath *)otherIndexPath
```
Comment (appleguy) [inline AsyncDisplayKit/Private/ASInternalHelpers.mm]: hmmm...in the layer extensions, we use asyncdisplaykit_ as the prefix.  I guess using asdk_ as @levi suggests would actually be better, but we should do a mass rename of the other properties in a later diff if so.

File: AsyncDisplayKit/Private/_ASHierarchyChangeSet.m [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    } else {
+      // Ending the current group
+      _ASHierarchyItemChange *change = [[_ASHierarchyItemChange alloc] initWithChangeType:type indexPaths:currentIndexPaths animationOptions:currentOptions presorted:YES];
+      [result addObject:change];
+      currentOptions = 0;
+      currentIndexPaths = nil;
+    }
+  }
+  
+  if (currentIndexPaths != nil) {
```
Comment (levi) [inline AsyncDisplayKit/Private/_ASHierarchyChangeSet.m]: This extra duplication step can be eliminated with a for loop

**Comment by levi** [comment]
Clean addition on top of data controller. Would love it if the reasoning behind the design was well documented at the `ASChangeSetDataController` level (as described in a massive conversation in #684) and @Adlai-Holler's design of the change set objects at the internal level. Thanks for doing this, @nguyenhuy , I can tell you there will be a few happy people on my team once this lands. :+1:

File: AsyncDisplayKit/Private/ASInternalHelpers.mm (hunk 2 above)
Comment (nguyenhuy) [inline AsyncDisplayKit/Private/ASInternalHelpers.mm]: Changed to asdk_ prefix.

**Comment by nguyenhuy** [comment]
@levi Thanks for taking a look. I agree with you that `ASChangeSetDataController` deserves a better documentation. I'm just not so sure whether people would agree with me about its existence. Its reasonings are separation of concerns, no UITableView/UICollectionView specific behaviours embedded in `ASDataController` and therefore reduce its complexity. Another minor benefit is that developers can switch back to `ASDataController` if `ASChangeSetDataController` causes any problems. I will add more documentation once others approve this approach.

That's said, it's really helpful if you could point out what are the things you would love to see in the documentation, from a new user (of that class) point of view.

File: AsyncDisplayKit/Private/_ASHierarchyChangeSet.m (hunk 3 above)
Comment (nguyenhuy) [inline AsyncDisplayKit/Private/_ASHierarchyChangeSet.m]: cc @Adlai-Holler

**Comment by appleguy** [comment]
@nguyenhuy can you look at the test failures with the build?  I did try re-running it, in case it was a flaky test, but it seems like it probably isn't.

**Comment by appleguy** [comment]
BTW - this is an extremely awesome patch and another huge contribution to the community.  I can't believe how fast you did this following @levi's diff landing, lol!!  Thank you!

**Comment by nguyenhuy** [comment]
Sure, I will look at the failed test. Looks like its the same one reported in #758.

Although I believe this diff should work nicely with @levi's supplementary-view feature, I admit that I haven't looked at the underlying diff closely and there might be some integration problems that I overlooked. I will recheck everything.

**Comment by appleguy** [comment]
Awesome!  @nguyenhuy once the test is fixed, do you consider this diff ready to land, or believe that any iteration is needed?

**Comment by nguyenhuy** [comment]
Probably one iteration needed. I will add short documentation and some tests, as well as fixing small implementation detials in `_ASHierarchyChangeSet`.

**Comment by levi** [comment]
@nguyenhuy just a quick glance at the conversations in the other issues, I'd love to see the documentation talk about the native ordering in UITableView/UICollectionView and that the `ASChangeSetDataController` is designed to simulate those characteristics. It's not entirely apparent to most people that table and collection views have a specific ordering of modification events.

**Comment by nguyenhuy** [comment]
@levi: yes, I will write that down. Thanks!

**Comment by Adlai-Holler** [comment]
Awesome PR! Pursuant to @levi's point, maybe include this link: https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/TableView_iPhone/ManageInsertDeleteRow/ManageInsertDeleteRow.html#//apple_ref/doc/uid/TP40007451-CH10-SW17

**Comment by nguyenhuy** [comment]
@levi I removed duplicate steps in _ASHierarchyChangeSet in #7e3df3b. Please help me to review it.

@levi, @Adlai-Holler: I also added documentation to ASChangeSetDataController. Let me know what you think.

@appleguy This PR is more or less ready to land. I will write some tests for these new classes in the next few days (and push to a new PR if this one landed by then).

**Comment by appleguy** [comment]
@nguyenhuy Thanks!  Do you think this is safe to include the 1.9 release if I were to push it today?  I would like to do the release today, and having this improvement would definitely be a milestone.

**Comment by nguyenhuy** [comment]
Yeah it would be nice to have this diff tested by other users. I will act quickly on any problems reported.

**Comment by appleguy** [comment]
@nguyenhuy could you take a quick look at the test failures now occurring on master?  I'm downloading Xcode 6.4 so I can try to reproduce locally, but it's definitely blocked my work for the weekend (wanted to get a lot of PRs merged today, but lots else to do, like writing release notes / preparing release / working on documentation / ASPagerNode etc)

**Comment by nguyenhuy** [comment]
Btw, the failed test looks familiar and should be resolved by #770.

**Comment by nguyenhuy** [comment]
Yeah, looking at the failures on master.

**Comment by appleguy** [comment]
@nguyenhuy it looks like master is green, but even re-running the tests here is failing.  It's possible the test is still flaky but I think this diff may have actually regressed something.

Check out the build...

Failures:
  0) -[ASTableViewTests testRelayoutAllRowsWithZeroSizeInitially] (AsyncDisplayKitTests.xctest)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
/Users/travis/build/facebook/AsyncDisplayKit/AsyncDisplayKitTests/ASTableViewTests.m:426: ((node.numberOfLayoutsOnMainThread) equal to (1)) failed: ("0") is not equal to ("1"):
423       for (int row = 0; row < NumberOfRowsPerSection; row++) {
424         NSIndexPath *indexPath = [NSIndexPath indexPathForRow:row inSection:section];
425         ASTestTextCellNode *node = (ASTestTextCellNode *)[tableView nodeForRowAtIndexPath:indexPath];
426         XCTAssertEqual(node.numberOfLayoutsOnMainThread, 1);
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
427         XCTAssertEqual(node.constrainedSizeForCalculatedLayout.max.width, newSize.width);
428       }
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
** TEST FAILED: 213 passed, 1 failed, 0 errored, 214 total ** (49140 ms)

**Comment by appleguy** [comment]
@nguyenhuy "fortunately", it looks like the test failures are real.  I re-confirmed master is passing tests on Xcode 6.4 locally, and checked out this PR to run them again and got the same failure as the builedbot (2 total tests in my case though, both testRelayoutAllRowsWithZeroSizeInitially and testRelayoutAllRowsWithNonZeroSizeInitially).

**Comment by nguyenhuy** [comment]
I think it's because Travis run the tests without virtually/temporary merging this branch into master, i.e the new fix on master isn't included. Let me rebase this branch.

**Comment by appleguy** [comment]
Oh, weird that it didn't rebase when re-running the build.  I'm pretty sure it normally does that, but maybe some other automatic action usually triggers the rebase builds.  Thanks, I'm excited to have almost all the open PRs merged.

Now we need master to be hit with a day of testing to pave out the recent table changes.  I'd like to reset the 1.9 release tag point to include the recent improvements, if at all possible.

Could folks bump the ASDK version you're testing against and let us know if you hit anything?  @rcancro, @eanagel, @levi, @Adlai-Holler

**Comment by appleguy** [comment]
@nguyenhuy hmm, still failed in the number of layouts on main thing...

https://travis-ci.org/facebook/AsyncDisplayKit/jobs/87395262

/Users/travis/build/facebook/AsyncDisplayKit/AsyncDisplayKitTests/ASTableViewTests.m:426: ((node.numberOfLayoutsOnMainThread) equal to (1)) failed: ("0") is not equal to ("1"):
423       for (int row = 0; row < NumberOfRowsPerSection; row++) {
424         NSIndexPath *indexPath = [NSIndexPath indexPathForRow:row inSection:section];
425         ASTestTextCellNode *node = (ASTestTextCellNode *)[tableView nodeForRowAtIndexPath:indexPath];
426         XCTAssertEqual(node.numberOfLayoutsOnMainThread, 1);
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
427         XCTAssertEqual(node.constrainedSizeForCalculatedLayout.max.width, newSize.width);
428       }

**Comment by nguyenhuy** [comment]
Ok, I will recheck everything here. Thanks for the heads up :)

**Comment by appleguy** [comment]
If it’s late wherever you are, don’t rush!  I need to write the release notes anyway, which I won’t have done for approximately 24hrs anyway.

		— Scott

**Comment by nguyenhuy** [comment]
It's early morning over here. But yeah, shouldn't rush this thing. Thanks.

**Comment by nguyenhuy** [comment]
@appleguy You are right, Travis does merge the branch with master and run the tests on top of the merge commit. I just had a hard time believing the tests were still failing and desperately looked for someone to blame! But I think I found out the root cause and fixed it. All checks have passed :clap:

**Comment by appleguy** [comment]
This may be one of the most satisfying-to-click Merge buttons I've done for ASDK :).  ASDataController has really gone from the Achilles heel of the framework to one of its most capable and unique components.

**Pull request merged**